### DISCUSSION TITLE
升级到 Vue3（旧标题：升级 Vue 版本到 2.7)

### DISCUSSION THREAD

**Discussion by timongh** [the1812/Bilibili-Evolved#3939] [功能建议 / Ideas]
Vue 2.7 是 Vue 2.x 的最后一个版本（支持到 2023 年 12 月 31 日），而且内置了组合式 API。可以考虑升级一下。

我自己先改改，改完了提个 PR 上来。

### see

[Vue 2 LTS， EOL & Extended Support](https://v2.cn.vuejs.org/lts/)

[Migration to Vue 2.7](https://v2.cn.vuejs.org/v2/guide/migration-vue-2-7.html)

**Comment by the1812** [comment]
我挺想直接升 3 的, 换组合式 API 总归要改代码

**Reply by timongh** [reply]
可以，我也挺想升 3 的。

现在的情况是，升级 Vue 需要更严格的 ts 类型检查（比如 noImplicitThis），另外类型检查工具我也换成了 vue-tsc。

只是升级 2.7 的话，我可以只改类型，有 bug 的代码我可以先用 `@ts-ignore` 忽略过去，而运行时的行为不会发生改变。升 3 可能就要连着类型和 bug 一起改，不然原来能跑的 bug 代码可能就不能跑了。这样一来，在所有代码改完之前，都不能合并到主分支里去。

还有一个问题是，核心升级后，原来用 Vue2 编写的组件可能就不能跑了，因此可能需要禁止加载老版本的组件，还要引导升级，这部分我不太熟，可能要你帮忙。

你看这样行吗，单独开个迁移 Vue3 的分支，我把已经改了的内容整理下发个 PR 到里面，然后我接着改。另外我再开个讨论，把我自己不好解决的问题写到那里去。

**Reply by timongh** [reply]
我先把一些已经发现的 BUG 作为 issue 提上来

**Reply by the1812** [reply]
> 还有一个问题是，核心升级后，原来用 Vue2 编写的组件可能就不能跑了，因此可能需要禁止加载老版本的组件，还要引导升级，这部分我不太熟，可能要你帮忙。

升 3 就是要另外搞个分支, 一次性弄好合到主分支才行, 所以升级完应该就不存在 Vue 2 的组件了. (我觉得主要的工作量也是用 3 的 API 重写这些组件)

**Reply by timongh** [reply]
不是啊，大部分代码还是可以重用的，Vue3 也可以用选项式的定义。只用把 Vue.extend 改成 defineComponent。再改改 2 跟 3 的破坏式更新就好了（虽然工作量也不小）。

**Reply by timongh** [reply]
> 所以升级完应该就不存在 Vue 2 的组件了

我的意思是，用户的油猴存储里存的代码还是老版本的代码，要禁止新的 core 加载旧的组件代码。因为我看过核心里有关版本的代码，隐约记得其中的逻辑好像是：如果找不到 version 这个字段，仍然会继续加载代码。

**Reply by the1812** [reply]
> 我的意思是，用户的油猴存储里存的代码还是老版本的代码，要禁止新的 core 加载旧的组件代码。因为我看过核心里有关版本的代码，隐约记得其中的逻辑好像是：如果找不到 version 这个字段，仍然会继续加载代码。

那确实, 之前认为旧组件代码也是兼容可用的, 应该得加个特殊处理了

**Comment by timongh** [comment]
src/ui/RangeInput.vue 能接受一个 `range`，类型是 `Range<?>`，但这个类型参数我从上下文推测不出来。只能看出他能调用 `toString()` 方法。

**Reply by timongh** [reply]
@the1812

**Reply by the1812** [reply]
在一旁的 `range.ts` 里
```ts
export interface Range<T = number> {
  start: T
  end: T
}
```

**Reply by timongh** [reply]
那这个 `T` 要限制为 `number` 吗？因为在上下文中对这个 `T` 调用了 `toString`，所以不能是 `unknown`，至少也应该是个 `{ toString: () => string }`。

**Reply by the1812** [reply]
~~目前在用的都是 `number`, 直接去掉泛型也行~~

**Reply by timongh** [reply]
不对，我之前没注意，这个地方只能是 `Range<string>`。因为 RangeInput 内部使用的是 TextBox，而 TextBox 的输入输出都只能是 `string`。

**Reply by the1812** [reply]
我记错了, 重新看了下代码, 确实是 string. (夜间模式计划时段有用到)

**Comment by timongh** [comment]
src/ui/VDropdown.vue 能接受一个 `value`，这个 `value` 的类型我也看不出来。只能从上下文看出，这个类型里面有个 `displayName` 属性。

**Reply by timongh** [reply]
@the1812

**Reply by the1812** [reply]
只是任意有这个属性的对象就行了, 但也只是默认在 slot 里读取 `displayName`, 这个 slot 可以被替换掉
```ts
interface Item {
  displayName: string
}
```

**Comment by timongh** [comment]
另外，目前大部分升级工作已经完成。只不过所有修改后的代码都还没有测试。

**Comment by timongh** [comment]
@the1812

目前不需要大改动的代码已经完成（逻辑改变不大的）。

剩下的因为涉及逻辑更改，至少需要被审核后才能合并，所以要提单独的 PR 才行。这部分我就等单独分支开出来再做了（因为要提交 PR）

**Comment by timongh** [comment]
@the1812 

目前的情况是：有一部分代码必须重构，否则无法正常工作。有部分代码推荐重构、优化。另外，版本的隔离和迁移没有做。所以现在是不能合并进 preview-features 的。

但是重构我认为至少还是要 review 一下。所以我的意见是建个  vue3 的分支。我通过提 PR 的方式来一个个地开发。

**Reply by the1812** [reply]
可以的, 我根据 `preview-features` 建一个?

**Reply by timongh** [reply]
恩恩

**Reply by the1812** [reply]
https://github.com/the1812/Bilibili-Evolved/tree/vue3

**Comment by timongh** [comment]
顺便问一下，如果不 @ 的话，不知道能否收到消息。

**Reply by the1812** [reply]
能的
